### DISCUSSION TITLE
Using user input outside of function (and throughout program)

### DISCUSSION THREAD

**Discussion by Geoff-McK** [jspsych/jsPsych#1056] [Q&A] [answered]
Hi everyone, 

I am programming an experiment that is a 2-day study. I have it set up so that on the second day, participants would type in their subject number, which would be used to get their condition information. I managed to get their user input to get their condition information. However, the variables that are created inside of the function do not seem to get used outside of the function. I understand that this is normal, but I still can't figure out how to use the input from the participant outside of the function. I tried adding these variables to the data (using jsPsych.addProperties) and then accessing them outside of the function (using jsPsych.data.get().values[0].subject, for example) but it didn't work.

Any assistance would be greatly appreciated. 

Here is the code that does this:

var survey_page2 = {
    type: 'html-keyboard-response',
    stimulus: '<p></p>',
    choices: jsPsych.NO_KEYS,
    trial_duration: 10,
    data: {task: 'data-input'},
       on_finish: function(data){
        // var A = parseInt(JSON.parse(jsPsych.data.get().filter({task: 'data-input'}).last(1).values()[0].responses).subject);
        console.log("subject:" + subject_info[A-1][0]);
        var interval = subject_info[A-1][1];
        console.log("interval:" + interval);
        var target = subject_info[A-1][2];
        console.log("target:" + target);
        jsPsych.data.addProperties({subject: A, interval: interval, target: target}); 

  } 
};

**Comment by becky-gilbert** [comment]
@Geoff-McK, just to check that I understand: participants enter their ID number at some point _before_ this `survey_page2` trial, and you add their response to the data with [`jsPsych.data.addProperties`](https://www.jspsych.org/core_library/jspsych-data/#jspsychdataaddproperties). Then on this `survey_page2` trial, it does nothing except use the `on_finish` function to access the participant ID and other info (since this trial has no valid key choices and only lasts 10ms). Is that right? Could you also share the code you're using to set up the ID question and save the response?

**Comment by Geoff-McK** [comment]
Hi Becky,

Thank you for your response. I have changed the program a bit so while the problem is fundamentally the same, the solution that I attempted in the following the code is different. I don't really have a preference for how to solve this problem, and this is just one of many ways that I approached it. I don't really even mind using plugins that may be better suited. As you'll see, I am really new at jsPsych and Javascript, and do not really have a grasp over the flow of my code. I will explain the study and code in detail. 

The study involves people watching a video of a mock crime. Immediately after the video, they are either asked to make a prediction about how well they think they would perform on future police lineup task or not. Upon returning a days later, the type in their subject number, and shown a police lineup and asked to select one of the photos if the person from the video is present, and reject the lineup if the person is not present. Half of the participants who did not make a judgment immediately after the video (on day 1) after asked to make a judgment prior to viewing the lineup. In the code below, I am only trying to a) get the subject number, b) assign the appropriate variables based on the subject number, and c) present a lineup to the subject. 

In the code below, the function called "survey_page2" asks the user for their subject number, using "survey-text" plugin. Using this number, information about which conditions they are in is collected. Lineups are assigned based on the values of video_order (i.e., video name) and target (i.e., whether the person from the video will be present or not). I also add these variables to the data. Note: this is different from my original post in which "survey_page2" used the input that was solicited in a separate function to assign variables to the user.

In the function called "lineup_trial", I present the photos from the "lineup" variable created in the above function in the preamble, using the "survey-html-form" plugin. I should mention that had to abbreviate the preamble in the "lineup_trial" function because the post looked strange after submitting it. 

var target = -1;
var interval = -1;
var lineup = [];

var survey_page2 = {
    type: 'survey-text',
    questions: [
        {prompt: "What is your subject number?", columns: 4, required: true, name: 'subject'}
    ],
    data: {task: 'data-input'},
       on_finish: function(data){

        var A = parseInt(JSON.parse(jsPsych.data.get().filter({task: 'data-input'}).last(1).values()[0].responses).subject);

        // subject_info is a matrix with 4 columns: subject, interval, target, and video_order
        var subject = subject_info[A-1][0];
        var interval = subject_info[A-1][1];
        var target = subject_info[A-1][2];
        var video_order = subject_info[A-1][3];
        
        if(video_order==0){
            if(target==0){
                lineup = ['CarTheftFiller1.jpg','CarTheftFiller2.jpg','CarTheftFiller3.jpg','CarTheftFiller4.jpg','CarTheftFiller5.jpg','CarTheftFiller6.jpg'];
            } else {
                lineup = ['CarTheftFiller1.jpg','CarTheftFiller2.jpg','CarTheftFiller3.jpg','CarTheftFiller4.jpg','CarTheftFiller5.jpg','CarTheftPerp.jpg'];
            }

        } else if (video_order==1){
            if(target===0){
                lineup = ['AutopartCrimeFiller3.jpg','AutopartCrimeFiller10.jpg','AutopartCrimeFiller11.jpg','AutopartCrimeFiller12.jpg','AutopartCrimeFiller14.jpg','AutopartCrimeFiller5.jpg'];
        } else {
                lineup = ['AutopartCrimeFiller3.jpg','AutopartCrimeFiller10.jpg','AutopartCrimeFiller11.jpg','AutopartCrimeFiller12.jpg','AutopartCrimeFiller14.jpg','AutopartCrimePerp.jpg'];
        }
        }

            // randomizes position of pictures in the lineup
            lineup = jsPsych.randomization.repeat(lineup, 1);

        // this is one way to add subject to data (I think)
        data.subject = parseInt(JSON.parse(jsPsych.data.get().filter({task: 'data-input'}).last(1).values()[0].responses).subject); 

        // this adds subject, interval, target, video_order, and lineup to the data
        jsPsych.data.addProperties({subject: subject, interval: interval, target: target, video_order: video_order, lineup: lineup});


  } 
}; 


var lineup_trial = {
		type: 'survey-html-form',
    preamble:
  "<div [...] + lineup[0] + [...]  " +  
  "<div [...] + lineup[1] + [...]  " + 
  "<div [...] + lineup[2] + [...]  " +
  "<div [...] + lineup[3] + [...]  " + 
  "<div [...] + lineup[4] + [...]  " + 
  "<div [...] + lineup[5] + [...]  ",

	html: "<div class='container' style = 'position: fixed; top: 530px; left: 375px;'> <p><br><br><input name='first' type='text' </p></div>" ,
	button_label: 'SUBMIT'

    }


  var timeline = [];
timeline.push(survey_page2)
timeline.push(lineup_trial);


jsPsych.init({
    show_preload_progress_bar: true,
    timeline: timeline,
    on_finish: function() {
    jsPsych.data.displayData();
   }
});

**Comment by becky-gilbert** [comment]
Ok thanks for clarifying and adding code - very helpful! If you're getting the subject number in the `on_finish` function for the trial where the subject enters their subject number, then you don't need to do all of the data filtering etc. because the `on_finish` function already has access to the data from that trial via the [function's 'data' argument.](https://www.jspsych.org/overview/callbacks/#on_finish-trial) 

So in the last version of the code that you posted, in the `on_finish` function for the `survey_page2` trial, you can just do this:
```javascript
// get the subject's response from this trial and convert it to an integer
var A = parseInt(JSON.parse(data.responses).subject);
```
The 'data' object that is passed into the `on_finish` function is modifiable, so you can also add the subject number to the data for that trial like this:
```javascript
var subject = subject_info[A-1][0]; // I can't test this line without the subject info matrix, but assuming it works...
data.subject = subject;
```
I think you have the right idea here when it comes to adding the subject number and other info to the data for _all_ trials in the experiment:
```javascript
// this adds subject, interval, target, video_order, and lineup to the data
jsPsych.data.addProperties({subject: subject, interval: interval, target: target, video_order: video_order, lineup: lineup});
```
If you want to access the subject number from somewhere other than the `on_finish` function for trial where the participant enters their number, for instance in the `on_finish` function for your `lineup_trial`, then you had the right idea about filtering the data on {task: 'data-input'}. Here's how to do it:
```javascript
var subject = jsPsych.data.get().filter({task: 'data-input'}).values()[0].subject;
```
And if you add the subject number to the data using `jsPsych.data.addProperties`, then that information will exist in _all_ of the trial data, which means you can always get it from the last trial data, like this:
```javascript
var subject = jsPsych.data.getLastTrialData().values()[0].subject;
```

**Comment by Geoff-McK** [comment, marked as answer]
I'm sorry, but I'm still a bit confused. I would like to access the lineup information in lineup_trial, so that I can present them. I believe that this means that getting the subject information (to get to the lineup information) during the on_finish function for lineup_trial. Basically, I'd like to do something like below; however, "data" isn't defined at the beginning of the study.

var lineup_trial = {
type: 'survey-html-form',
preamble:
"<div [...] + data.lineup[0] + [...] " +
"<div [...] + data.lineup[1] + [...] " +
"<div [...] + data.lineup[2] + [...] " +
"<div [...] + data.lineup[3] + [...] " +
"<div [...] + data.lineup[4] + [...] " +
"<div [...] + data.lineup[5] + [...] ", 


}

**Reply by becky-gilbert** [reply]
Ah ok, I see what you mean now. You're right, the trial data isn't defined inside of the preamble parameter. Instead, you'll want to make the `preamble` parameter for your `lineup_trial` into a function, since you need to determine the specific value of this parameter _during_ your experiment, after you know the subject number and set the `lineup` array (rather than when the page first loads). See the docs on [dynamic parameters](https://www.jspsych.org/overview/trial/#dynamic-parameters).
```javascript
var lineup_trial = {
type: 'survey-html-form',
preamble: function() {
    var preamble_html = "<div [...] + lineup[0] + [...] " +
    "<div [...] + lineup[1] + [...] " +
    "<div [...] + lineup[2] + [...] " +
    "<div [...] + lineup[3] + [...] " +
    "<div [...] + lineup[4] + [...] " +
    "<div [...] + lineup[5] + [...] ";
  return preamble_html;
}
```
Does that work?

**Reply by Geoff-McK** [reply]
Yes! It worked!!! Thank you!
